dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1199 @@
+# -*- coding: utf-8 -*-
+"""Creates a GCP cloud resource tree."""
+
+import json
+import tempfile
+from datetime import datetime
+from datetime import timedelta
+from datetime import timezone
+from typing import Dict, List, Optional, Any, Set
+from enum import Enum, auto
+
+from dateutil import parser
+from libcloudforensics.providers.gcp.internal.common import CreateService
+from libcloudforensics.providers.gcp.internal.log import GoogleCloudLog
+
+from dftimewolf.lib import module
+from dftimewolf.lib.containers import containers
+from dftimewolf.lib.modules import manager as modules_manager
+from dftimewolf.lib.state import DFTimewolfState
+
+
+class OperatingMode(Enum):
+  """Enum represent operational mode (Online or Offline)."""
+
+  ONLINE = auto()
+  OFFLINE = auto()
+
+
+class Resource:
+  """A Class that represent a resource (Instance, Disk, Image...etc)."""
+
+  has_dynamic_attributes = True  # silences all attribute-errors for Resource
+
+  def __init__(self) -> None:
+    """Initialize the Resource object."""
+    self.id = ''
+    self.name = ''
+
+    self.type = ''
+    self.state = ''
+    self.project_id = ''
+    self.zone = ''
+
+    self.created_by = ''
+    self.creator_ip_address = ''
+    self.creator_useragent = ''
+    self.deleted_by = ''
+    self.deleter_ip_address = ''
+    self.deleter_useragent = ''
+    self.parent: Optional[Resource] = None
+    self.children: Set[Resource] = set()
+    self.disks: List[Resource] = []
+    self.deleted = False
+    self._resource_name = ''
+    self._creation_timestamp: Optional[datetime] = None
+    self._deletion_timestamp: Optional[datetime] = None
+
+  def __hash__(self) -> int:
+    """For object comparison."""
+    return hash(self.id + self.resource_name)
+
+  @property
+  def resource_name(self) -> str:
+    """Property resource_name."""
+    return self._resource_name
+
+  @resource_name.setter
+  def resource_name(self, value: str) -> None:
+    """Property resource_name Setter."""
+    if value:
+      values = value.split('/')
+
+      self.name = values[-1]
+
+      resource_type = values[-2]
+      if resource_type == 'disks':
+        self.type = 'gce_disk'
+      elif resource_type == 'instances':
+        self.type = 'gce_instance'
+      elif resource_type == 'images':
+        self.type = 'gce_image'
+      elif resource_type == 'machineImages':
+        self.type = 'gce_machine_image'
+      elif resource_type == 'instanceTemplates':
+        self.type = 'gce_instance_template'
+      elif resource_type == 'snapshots':
+        self.type = 'gce_snapshot'
+      else:
+        self.type = resource_type
```

Comment:
what is the purpose of matching the default resource_type value from logs to new custom values?
```suggestion
      self.type = values[-2]
```

Reply:
The type will be used in the command line input by the user where they supply the resource name and resource type. I created those to make the type values more user friendly. so instead of supplying "dm2 machineImages" in the argument, the user can supply "dm2 gcp_machine_image". The camelcase is not intuitive and having the plural of the type doesn't make sense for the user.